draco/core/misc.py

```diff
@@ -214,6 +217,50 @@ def finish(self):
         return items
 
 
+class CheckMPIEnvironment(task.MPILoggedTask):
+    """Check that the current MPI environment can communicate
+    across all nodes.
+    """
+
+    timeout = config.Property(proptype=int, default=240)
+
+    def setup(self):
+        from caput import mpiutil
+        import time
+
+        comm = self.comm
+        n = 500000  # Corresponds to about a 4 mb buffer
+        results = []
+
+        sends = np.arange(comm.size * n, dtype=float).reshape(comm.size, n)
+        recvs = np.empty_like(sends)
+
+        def prep_buf(x):
+            return mpiutil.typemap(x.dtype)
+
+        # Send and receive across all ranks
+        for i in range(comm.size):
+            send = (comm.rank + i) % comm.size
+            recv = (comm.rank - i) % comm.size
+
+            results.append(comm.Irecv(prep_buf(recvs[recv, :]), recv))
+            comm.Isend(prep_buf(sends[comm.rank, :]), send)
```

Comment:
I don't think you need a `prep_buf` in here. mpi4py should figure it out automatically
```suggestion
        # Send and receive across all ranks
        for i in range(comm.size):
            send = (comm.rank + i) % comm.size
            recv = (comm.rank - i) % comm.size

            results.append(comm.Irecv(recvs[recv, :], recv))
            comm.Isend(sends[comm.rank, :], send)
```

Reply:
Ah, yeah, I had needed it in a previous iteration but should be able to remove it now

Reply:
You might also want to give an explicit width type for the numpy array. It's easier to reason about. I think `float` -> `np.float64` but I'm never entirely sure.

Reply:
Fair